[Anunaya Joshi] IP

anunayajoshi opened the pull request:
### This PR consists of changes to include **CRUD** functionality and _Unit Testing_ to the application. 😄 

The changes made are as follows:

- Ability to greet and exit 
- Ability to add, list, delete todos, deadlines and events 
- Automated Text UI Testing as well as JUnit
- Saving to hard disk
- Marking and unmarking tasks as done or not.

The things I struggled with while working on this application were:

1. Following OOP best practices and deciding on the classes to use. 
2. Using git and the branches effectively and dealing with merge conflicts. 

Additional things that need to be done are:

- [ ] Creating a GUI for the application
- [ ] Adding in checkstyle functionality 

If you would like to view work done by me, you can visit by github [here](www.github.com/anunayajoshi)

One quote I like is

> Spend more time and do it once, instead of doing it multiple times. 


A sample input for the application is like `mark 3` 

Example code function is as follows:

```java
public void setTaskStatus(int index, boolean isDone) {
        Task task = this.tasks.get(index - 1);
        task.setDone(isDone);
        Ui.separator();
        System.out.println("\tOk, I have marked this task as " + (isDone ? "done" : "not done yet") + ":\n\t\t"
                + task);
        Ui.separator();
    }


```

Commit:
Add Gradle support

anunayajoshi committed:
Level-1, included Greet, Echo and Exit

anunayajoshi committed:
Level 2: Add, list

anunayajoshi committed:
Mark as Done and A-Classes

anunayajoshi committed:
Add Todos, Events, Deadlines

anunayajoshi committed:
Automated Text UI Testing

anunayajoshi committed:
Handle errors and A-Exceptions

anunayajoshi committed:
Delete

anunayajoshi committed:
Contain Error messages for events and deadlines

anunayajoshi committed:
Include unit tests for deadline and lists

anunayajoshi committed:
Merge branch 'master' into feature1

anunayajoshi committed:
Include more unit tests

anunayajoshi committed:
Save data to hard disk

anunayajoshi committed:
Include Dates for deadlines

anunayajoshi committed:
Solving merge conflicts

anunayajoshi committed:
Fix bug of duke.txt date type

anunayajoshi committed:
Merge remote-tracking branch 'origin/add-gradle-support'

anunayajoshi committed:
Use Gradle and fix duke.txt LocalDate parsing

anunayajoshi committed:
Tasklist and Storage OOP Improvements

anunayajoshi committed:
A-MoreOOP

anunayajoshi committed:
Dividing classes into packages

anunayajoshi committed:
Add J-Unit Tests

anunayajoshi committed:
Jar File

anunayajoshi committed:
Add JavaDoc

anunayajoshi committed:
A-CodingStandard

anunayajoshi committed:
Find

anunayajoshi committed:
Merge branch 'branch-A-CodingStandard'

anunayajoshi committed:
Resolving merge conflict

anunayajoshi committed:
Update Javadocs

anunayajoshi committed:
Merge branch 'branch-A-JavaDoc'

anunayajoshi committed:
Add Checkstyle Config

anunayajoshi committed:
Merge branch 'branch-A-CheckStyle'

anunayajoshi committed:
Provide GUI using JAVAFx

anunayajoshi committed:
Merge branch 'branch-Level-10'

huixuant commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    /**
+     * Takes in user's input and obtains the first word
```
Perhaps the format of the JavaDoc header comment here can be further improved? The coding standard states that the first sentence should be written as a **short summary of the method**, as Javadoc automatically places it in the method summary table (and index). However, the first line of this header comment appears to be truncated and not complete. 

Additionally, perhaps consider adding an empty line between the description and parameter section, as per the coding standard? This applies to all other JavaDoc comments without the empty line.

huixuant commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,107 @@
+        String str;
+        try {
+            String[] arr = input.split(" ", 2);
+            boolean details = arr.length != 1;
```
Perhaps consider using another name for the boolean variable `details`? According to the coding standard, boolean variables should be named to sound like booleans. A possible alternative could be something like `hasDetails`. 😄

huixuant commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class Task {
+    private final String description;
+    private Boolean done;
```
Similar to one of the comments above, perhaps consider renaming `done` to `isDone` so that it complies with the coding standard?

huixuant commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,51 @@
+        return sc.hasNextLine();
+    }
+
+    public String userInput() {
```
Perhaps consider renaming this method to start with a verb so that it complies with the coding standard? For example, a possible alternative could be `getUserInput()`.

huixuant commented on src/test/java/duke/EventTest.java:
```diff
@@ -0,0 +1,24 @@
+     * Tests if the event toString method prints correctly.
+     */
+    @Test
+    public void eventToStringTest(){
```
Perhaps the test method could be renamed such that it follows the format given in the coding standard `featureUnderTest_testScenario_expectedBehavior()`?

huixuant reviewed (commented):
Code seems well-organised with elements of OOP being implemented. Other than minor coding standard violations, code is generally readable and easy to understand with proper naming in place.

MrTwit99 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+     * Creates a Deadline Task
+     *
+     * @param description in String format.
+     * @param by          Which is a LocalDate.
```
Is there perhaps some additional space indented error or there was something here but I can't view it?
```suggestion
     * @param by Which is a LocalDate.
```

MrTwit99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,107 @@
+            storage.writeFile(tasklist);
+        } catch (DukeException e) {
+            str = ui.showError(e);
+
```
There might be an additional space that was accidentally entered. Not big of an issue.

```suggestion
```

MrTwit99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,91 @@
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.Scanner;
+
```
I was just wondering, would it be slightly better to leave a line to separate the imports from diff packages iirc 2030S used to do this... Not sure if its needed here.

```suggestion
import java.io.IOException;

import java.util.Scanner;

```

MrTwit99 replied:
Edit to above comment:

After running the checkstyle feature, I think the above suggestion is invalid for most ``import`` situations! Do take note when there's a need for it.

MrTwit99 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,100 @@
+public class TaskList {
+
+    private final ArrayList<Task> tasks;
+
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
+
+    public ArrayList<Task> getTasks() {
+        return this.tasks;
+    }
+
+    public String list() {
+        String list = "";
+        for (int i = 0; i < this.tasks.size(); i++) {
+            int currentNumber = i + 1;
+            Task task = this.tasks.get(i);
+
+            list = list + "\t" + currentNumber + "." + task + "\n";
+        }
+        return list;
+    }
+
```
Would be great if you could add the JavaDocs in when you have the time.

MrTwit99 reviewed (commented):
Overall, there isn't much issues besides some naming issues as mentioned by another reviewer. Great job with the iP project and you could perhaps review through the suggestions I have made.

anunayajoshi committed:
Use Assertions

anunayajoshi committed:
Improve code quality

anunayajoshi committed:
Merge pull request #2 from anunayajoshi/branch-A-Assertions

branch-A-Assertions

anunayajoshi committed:
Merge branch 'master' into branch-A-CodeQuality

anunayajoshi committed:
Merge pull request #3 from anunayajoshi/branch-A-CodeQuality

Improve code quality

anunayajoshi committed:
Add Friendlier Syntax

anunayajoshi committed:
Update UI Design to add padding

Included padding to the left of the Image dialogBox to make it look
cleaner.

anunayajoshi committed:
C-FlexibleSyntax

anunayajoshi committed:
Fix DukeException Error Printing

There was an error of the error message being printed together with
duke.dukeException.

What had to be done was to use err.getMessage() to solve the issue

anunayajoshi committed:
Better GUI

anunayajoshi committed:
Personality

anunayajoshi committed:
Add Ui.png

anunayajoshi committed:
Add User Guide

anunayajoshi committed:
Fix Table Format

anunayajoshi committed:
Add new line spacing

anunayajoshi committed:
Fix checkStyle
